src/transformStateWithClones.js

```diff
@@ -8,6 +8,31 @@
  */
 function transformStateWithClones(state, actions) {
   // write code here
+  let currentState = state;
+  const stateHistory = [];
+
+  for (let i = 0; i < actions.length; i++) {
+    const action = actions[i];
+
+    if (action.type === 'clear') {
+      currentState = {};
+    } else if (action.type === 'addProperties') {
+      currentState = { ...currentState, ...action.extraData };
+    } else if (action.type === 'removeProperties') {
+      const newState = { ...currentState };
+
+      for (let j = 0; j < action.keysToRemove.length; j++) {
+        const key = action.keysToRemove[j];
+
+        delete newState[key];
+      }
+      currentState = newState;
+    }
+
+    stateHistory.push(currentState);
+  }
+
+  return stateHistory;
```

Comment:
Consider using a `switch` statement instead of multiple `if-else` conditions for better readability and maintainability, especially since you're dealing with a limited set of action types.